### PR TITLE
ARROW-9475: [Java] Clean up usages of BaseAllocator, use BufferAllocator in…

### PR DESCRIPTION
**Pull request by zhztheplayer**
…stead

Issue link: https://issues.apache.org/jira/browse/ARROW-9475.

### CODE REVIEW

**Comment by zhztheplayer** [comment]
This patch is also supposed to be a dependency of #7030 ([ARROW-7808](https://issues.apache.org/jira/browse/ARROW-7808)).

File: java/memory/memory-core/src/main/java/org/apache/arrow/memory/BufferAllocator.java [hunk 1]
```diff
@@ -126,6 +134,30 @@ BufferAllocator newChildAllocator(
    */
   long getHeadroom();
 
+  /**
+   * Forcibly allocate bytes. Returns whether the allocation fit within limits.
+   *
+   * @param size to increase
+   * @return Whether the allocation fit within limits.
+   */
+  boolean forceAllocate(long size);
```
Comment (rymurr) [inline java/memory/memory-core/src/main/java/org/apache/arrow/memory/BufferAllocator.java:143]: I didn't see where this or releaseBytes are used in code. What is the purpose of releaseBytes and of makign forceAllocate public

**Review by rymurr — COMMENTED** [review]
Looks good to me, one comment about the allocate/release additions. Otherwise its great to further limit BaseAllocators visibility

File: java/memory/memory-core/src/main/java/org/apache/arrow/memory/BufferAllocator.java (hunk 1 above)
Comment (zhztheplayer) [reply java/memory/memory-core/src/main/java/org/apache/arrow/memory/BufferAllocator.java:143]: Hi @rymurr,

E.g. In AllocationMangager code, we were calling BaseAllocator#releaseBytes:

https://github.com/apache/arrow/blob/d9525f1b19c00e440b3af83a68d1f2bba89a0a24/java/memory/memory-core/src/main/java/org/apache/arrow/memory/AllocationManager.java#L155

After changes in this PR, we are now trying to call BufferAllocator#releaseBytes:

https://github.com/apache/arrow/pull/7768/files#diff-68d17bac8aafb7004e2fb310f131df5fR160

That's why the method is added to BufferAllocator.

**Comment by zhztheplayer** [comment]
@jacques-n Could you please talk a look? Thanks. 

For ARROW-7808, having this one solved we could be able to leave customized AllocationManager implementation out of the package `org.apache.arrow.memory`.

File: java/memory/memory-core/src/main/java/org/apache/arrow/memory/BufferAllocator.java (hunk 1 above)
Comment (rymurr) [reply java/memory/memory-core/src/main/java/org/apache/arrow/memory/BufferAllocator.java:143]: :+1:  sorry, missed that

**Review by rymurr — APPROVED** [review]

**Comment by emkornfield** [comment]
+1

**Pull request closed without merging**